src/mlpack/methods/preprocess/preprocess_string_main.cpp

```diff
@@ -0,0 +1,301 @@
+    PRINT_CALL("preprocess_string", "actual_dataset", "X",
+        "preprocess_dataset", "X", "lowercase", 1, "stopwords",
+        1, "punctuation", 1, "dimension", 3) +
+    "\n\n",
+    SEE_ALSO("@preprocess_binarize", "#preprocess_binarize"),
+    SEE_ALSO("@preprocess_describe", "#preprocess_describe"),
+    SEE_ALSO("@preprocess_imputer", "#preprocess_imputer"));
+
+PARAM_STRING_IN_REQ("actual_dataset", "File containing the reference dataset.",
+    "t");
+PARAM_STRING_IN_REQ("preprocess_dataset", "File containing the preprocess "
+    "dataset.", "o");
+PARAM_STRING_IN("column_delimiter", "delimeter used to seperate Column in files"
+    "example '\\t' for '.tsv' and ',' for '.csv'.", "d", "\t");
+PARAM_STRING_IN("delimiter", "A set of chars that is used as delimeter to"
+    "tokenize the string dataset ", "D", " ");
+PARAM_STRING_IN("stopwordsfile", "File containing stopwords", "S", "");
+
+PARAM_FLAG("lowercase", "convert to lowercase.", "l");
+PARAM_FLAG("punctuation", "Remove punctuation.", "p");
+PARAM_FLAG("stopwords", "Remove stopwords.", "s");
+PARAM_VECTOR_IN_REQ(std::string, "dimension", "Column which contains the"
+    "string data. (1 by default)", "c");
+
+using namespace mlpack;
+using namespace mlpack::util;
+using namespace arma;
+using namespace std;
+
+void createDataset(std::vector<std::vector<std::string>>& dataset,
+                   const std::string& filename,
+                   const std::string& column_delimiter)
+{
+  // Extracting the Contents of file
+  // File pointer
+  ifstream fin;
+  // Open an existing file
+  fin.open(filename);
+  if (!fin.is_open())
+  {
+    Log::Fatal << "Unable to open input file \n";
+  }
+  std::string line, word;
+  boost::string_view token, copy;
+  while (std::getline(fin, line))
+  {
+    stringstream streamLine(line);
+    dataset.emplace_back(std::vector<std::string>());
+    // delimeter[0] becase the standard function accepts char as input.
+    while (std::getline(streamLine, word, column_delimiter[0]))
+    {
+      dataset.back().push_back(word);
+    }
+  }
+}
+
+void checkForDigit(const std::string& column)
+{
+  for (auto& i : column)
+    if (!std::isdigit(i))
+    Log::Fatal << "Characters encountered, Please give only digits";
+}
+
+void getColumn(const std::vector<std::string>& temp_dimension,
+               std::set<size_t>& dimension)
+{
+  int columnstartindex, columnendindex;
+  size_t found;
+  for (size_t i = 0; i < temp_dimension.size(); i++)
+  {
+    found = temp_dimension[i].find('-');
+    if ( found != string::npos)
+    {
+      try
+      {
+        // Has a range include, something of type a-b.
+        checkForDigit(temp_dimension[i].substr(0, found));
+        checkForDigit(temp_dimension[i].substr(found+1,
+            temp_dimension[i].length()));        
+        columnstartindex = std::stoi(temp_dimension[i].substr(0, found));
+        columnendindex = std::stoi(temp_dimension[i].substr(found+1,
+            temp_dimension[i].length()));
```

Comment:
You use `substr()` two times in order to get the same substring. Perhaps the compiler can optimize it out. I suggest to save the result to a variable.